### PR TITLE
Add partial support for XEP-0249 Direct MUC Invitations

### PR DESCRIPTION
**Pull request by MF1-MS**
This MR adds in partial support for [XEP-0249 Direct MUC Invitations](https://xmpp.org/extensions/xep-0249.html)

It exposes a method for a MUC to invite a user to the room, and adds a listener to the MultiUserChat listener to inform users of direct invitations they have received. 

A new UT class is added to test this function.

I found I could not build Smack without updating bouncy castle to 1.71 so have done so in this MR. Please advise how I would build without doing this if this is not desired.

### CODE REVIEW

File: smack-extensions/src/main/java/org/jivesoftware/smackx/muc/MultiUserChatManager.java [hunk 1]
```diff
@@ -254,6 +264,44 @@ public void run() {
...
+        StanzaListener xep0249InvitationPacketListener = new StanzaListener() {
+            @Override
+            public void processStanza(Stanza packet) {
+                final Message message = (Message) packet;
+                ExtensionElement inviteExt =
+                    packet.getExtension(GroupChatInvitation.class);
+                GroupChatInvitation invite =
+                    inviteExt instanceof GroupChatInvitation ? (GroupChatInvitation) inviteExt : null;
+
+                if (invite != null && invite.getRoomAddress() != null) {
```
Comment (Flowdalic) [inline smack-extensions/src/main/java/org/jivesoftware/smackx/muc/MultiUserChatManager.java:279]: Those checks are not needed. The `invite.getRoomAddress() != null` check is unnecessary,because the provider will not create GroupChatInviation instances without a room address, because the according element is required to have a 'jid' attribute.

**Review by Flowdalic — CHANGES_REQUESTED** [review]
Thanks for your contribution. I have left some remarks inline.

I think the most important issue that this is currently somehow entangled with some arbitrary `MultiUserChat` instance, when its business logic should really go into a new class `DirectMucInviteManager`.

**Comment by Flowdalic** [comment]
> I found I could not build Smack without updating bouncy castle to 1.71 so have done so in this MR. Please advise how I would build without doing this if this is not desired.

Bumping BC is fine. Since you give not details why you couldn't build without updating, I can't really help you there.

File: build.gradle [hunk 2]
```diff
@@ -148,7 +148,7 @@ allprojects {
 		smackMinAndroidSdk = 19
 		junitVersion = '5.7.1'
 		commonsIoVersion = '2.6'
-		bouncyCastleVersion = '1.70'
+		bouncyCastleVersion = '1.71'
```
[the remark this answers is not in the document]
Comment (MF1-MS) [reply build.gradle:151]: When I try to build/test Smack at 1.70 at get the following stacktrace. The method it is trying to access was introduced in BC 1.71. 

```
Starting a Gradle Daemon, 1 incompatible Daemon could not be reused, use --status for details

> Task :smack-openpgp:test

SecretKeyBackupHelperTest > createAndDecryptSecretKeyElementTest() FAILED
    java.lang.NoSuchMethodError: 'org.bouncycastle.asn1.x9.X9ECParametersHolder org.bouncycastle.asn1.x9.ECNamedCurveTable.getByOIDLazy(org.bouncycastle.asn1.ASN1ObjectIdentifier)'
        at org.bouncycastle.openpgp.PGPPublicKey.init(Unknown Source)
        at org.bouncycastle.openpgp.PGPPublicKey.<init>(Unknown Source)
        at org.bouncycastle.openpgp.PGPSecretKeyRing.<init>(Unknown Source)
        at org.bouncycastle.openpgp.PGPObjectFactory.nextObject(Unknown Source)
        at org.pgpainless.key.parsing.KeyRingReader.readSecretKeyRingCollection(KeyRingReader.java:246)
        at org.pgpainless.key.parsing.KeyRingReader.readSecretKeyRingCollection(KeyRingReader.java:221)
        at org.pgpainless.key.parsing.KeyRingReader.secretKeyRingCollection(KeyRingReader.java:74)
        at org.jivesoftware.smackx.ox.store.filebased.FileBasedOpenPgpKeyStore.readSecretKeysOf(FileBasedOpenPgpKeyStore.java:124)
        at org.jivesoftware.smackx.ox.store.abstr.AbstractOpenPgpKeyStore.getSecretKeysOf(AbstractOpenPgpKeyStore.java:146)
        at org.jivesoftware.smackx.ox.store.abstr.AbstractOpenPgpKeyStore.importSecretKey(AbstractOpenPgpKeyStore.java:162)
        at org.jivesoftware.smackx.ox.store.abstr.AbstractOpenPgpStore.importSecretKey(AbstractOpenPgpStore.java:131)
        at org.jivesoftware.smackx.ox.SecretKeyBackupHelperTest.createAndDecryptSecretKeyElementTest(SecretKeyBackupHelperTest.java:83)

29 tests completed, 1 failed
```

File: smack-extensions/src/main/java/org/jivesoftware/smackx/muc/MultiUserChatManager.java [hunk 3]
```diff
@@ -139,6 +142,11 @@ public static synchronized MultiUserChatManager getInstanceFor(XMPPConnection co
     private static final StanzaFilter INVITATION_FILTER = new AndFilter(StanzaTypeFilter.MESSAGE, new StanzaExtensionFilter(new MUCUser()),
                     new NotFilter(MessageTypeFilter.ERROR));
 
+    private static final StanzaFilter XEP_0249_INVITATION_FILTER =
+        new AndFilter(StanzaTypeFilter.MESSAGE,
+                      new StanzaExtensionFilter(GroupChatInvitation.ELEMENT, GroupChatInvitation.NAMESPACE),
```
[the remark this answers is not in the document]
Comment (MF1-MS) [reply smack-extensions/src/main/java/org/jivesoftware/smackx/muc/MultiUserChatManager.java:147]: `ExtensionElementFilter` has a protected constructer and won't let me call it from here. Should it be public?

File: smack-extensions/src/main/java/org/jivesoftware/smackx/muc/MultiUserChatManager.java [hunk 4]
```diff
@@ -199,6 +207,8 @@ public void processStanza(Stanza packet) {
         };
         connection.addAsyncStanzaListener(invitationPacketListener, INVITATION_FILTER);
 
+        addXEP0249Listener(connection);
```
[the remark this answers is not in the document]
Comment (MF1-MS) [reply smack-extensions/src/main/java/org/jivesoftware/smackx/muc/MultiUserChatManager.java:210]: The machinery to listen for invitations is not part of a `MultiUserChat` class, but part of the singleton `MultiUserChatManager` that manages a connection's MUCs. It has responsibility to listen for invitations and kick off having the desired XMPP client entity join the specified MUC room. 

The logic on a `MultiUserChat` instance scope is the ability to invite a client entity directly to that room, which also seems logical valid to me?

File: smack-extensions/src/main/java/org/jivesoftware/smackx/muc/MultiUserChat.java [hunk 5]
```diff
@@ -1063,6 +1065,29 @@ public void invite(MessageBuilder messageBuilder, EntityBareJid user, String rea
...
+     * Invites another user to the room in which one is an occupant. In contrast
+     * to the method "invite", the invitation is sent directly to the user rather
+     * than via the chat room.  This is useful when the user being invited is
+     * offline, as otherwise the invitation would be dropped.
+     *
+     * @param user the user to send the invitation to
+     * @throws NotConnectedException if the XMPP connection is not connected.
+     * @throws InterruptedException if the calling thread was interrupted.
+     */
+    public void inviteDirectly(EntityBareJid user) throws NotConnectedException, InterruptedException {
```
[the remark this answers is not in the document]
Comment (MF1-MS) [reply smack-extensions/src/main/java/org/jivesoftware/smackx/muc/MultiUserChat.java:1078]: True, I shall try to add soon.

File: smack-extensions/src/main/java/org/jivesoftware/smackx/muc/MultiUserChatManager.java (hunk 3 above)
[the remark this answers is not in the document]
Comment (Flowdalic) [reply smack-extensions/src/main/java/org/jivesoftware/smackx/muc/MultiUserChatManager.java:146]: It should be, fixed in master.

[the remark this answers is not in the document]
Comment (Flowdalic) [reply smack-extensions/src/main/java/org/jivesoftware/smackx/muc/MultiUserChatManager.java:146]: It should, fixed in master.

File: smack-extensions/src/main/java/org/jivesoftware/smackx/muc/MultiUserChatManager.java [hunk 6]
```diff
@@ -199,6 +206,40 @@ public void processStanza(Stanza packet) {
...
+        // Listens for all messages that include an XEP-0249 GroupChatInvitation extension and fire the invitation
+        // listeners
+        StanzaListener directInvitationPacketListener = new StanzaListener() {
+            @Override
+            public void processStanza(Stanza packet) {
+                final Message message = (Message) packet;
+                GroupChatInvitation invite =
+                    packet.getExtension(GroupChatInvitation.class);
+
+                if (invite.getRoomAddress() != null) {
```
Comment (Flowdalic) [inline smack-extensions/src/main/java/org/jivesoftware/smackx/muc/MultiUserChatManager.java:218]: This check is unnecessary, as it is impossible to create a `GroupChatInvitation` that hasn't set a room address.

File: smack-extensions/src/main/java/org/jivesoftware/smackx/muc/MultiUserChatManager.java [hunk 7]
```diff
@@ -199,6 +206,40 @@ public void processStanza(Stanza packet) {
...
+                    EntityBareJid mucJid = JidCreate.entityBareFromOrNull(invite.getRoomAddress());
+                    if (mucJid == null) {
```
Comment (Flowdalic) [inline smack-extensions/src/main/java/org/jivesoftware/smackx/muc/MultiUserChatManager.java:220]: Those checks are now also unnecessary, as the room address is always an `EntityBareJid`.

File: smack-extensions/src/main/java/org/jivesoftware/smackx/muc/MultiUserChatManager.java [hunk 8]
```diff
@@ -199,6 +206,40 @@ public void processStanza(Stanza packet) {
...
+                        LOGGER.warning("Received invite to non bare JID: '" + message + "'");
+                        return;
+                    }
+                    // Fire event for invitation listeners
+                    final MultiUserChat muc = getMultiUserChat(mucJid);
+                    final XMPPConnection connection = connection();
+                    final MUCUser.Invite mucInvite = null;
+                    final EntityJid from = message.getFrom().asEntityJidIfPossible();
+                    if (from == null) {
+                        LOGGER.warning("Group Chat Invitation from non entity JID: '" + message.toXML() + "'");
```
Comment (Flowdalic) [inline smack-extensions/src/main/java/org/jivesoftware/smackx/muc/MultiUserChatManager.java:230]: I am not super happy about that, as I believe it is likely that components, i.e., domain bare JIDs, could also send directed invites. I think we could change the signature of `InvitationsLIstener` to use just `Jid` for `inviter`.

File: smack-extensions/src/main/java/org/jivesoftware/smackx/muc/MultiUserChatManager.java [hunk 9]
```diff
@@ -199,6 +206,40 @@ public void processStanza(Stanza packet) {
...
+                if (invite.getRoomAddress() != null) {
+                    EntityBareJid mucJid = JidCreate.entityBareFromOrNull(invite.getRoomAddress());
+                    if (mucJid == null) {
+                        LOGGER.warning("Received invite to non bare JID: '" + message + "'");
+                        return;
+                    }
+                    // Fire event for invitation listeners
+                    final MultiUserChat muc = getMultiUserChat(mucJid);
+                    final XMPPConnection connection = connection();
+                    final MUCUser.Invite mucInvite = null;
```
Comment (Flowdalic) [inline smack-extensions/src/main/java/org/jivesoftware/smackx/muc/MultiUserChatManager.java:227]: Now that is a problem: `invite` was previously always not null. This has the potential to break apps if they upgrade to a newer Smack version. I think you want to synthesize a `MUCUser.Invite` here.

File: smack-extensions/src/main/java/org/jivesoftware/smackx/muc/MultiUserChatManager.java [hunk 10]
```diff
@@ -199,6 +206,40 @@ public void processStanza(Stanza packet) {
...
+                    final EntityJid from = message.getFrom().asEntityJidIfPossible();
+                    if (from == null) {
+                        LOGGER.warning("Group Chat Invitation from non entity JID: '" + message.toXML() + "'");
+                        return;
+                    }
+                    final String reason = null;
```
Comment (Flowdalic) [inline smack-extensions/src/main/java/org/jivesoftware/smackx/muc/MultiUserChatManager.java:233]: You should obtain `reason` and `password` from the xep249 directed user invite.

File: smack-extensions/src/test/java/org/jivesoftware/smackx/muc/MultiUserChatTest.java [hunk 11]
```diff
@@ -0,0 +1,211 @@
...
+    }
+
+    @Test
+    public void shouldReceiveOfflineInvitation() throws XmppStringprepException {
+        EntityBareJid roomJid = JidCreate.entityBareFrom("room@example.com");
+        EntityFullJid inviterJid = JidCreate.entityFullFrom("inviter@example.com/user1");
+        EntityBareJid inviteeJid = JidCreate.entityBareFrom("invitee@example.com");
+
+        GroupChatInvitation groupChatInvitation = new GroupChatInvitation(JidCreate.entityBareFrom("room@example.com"));
+        Message sentMessage = MessageBuilder.buildMessage(UUID.randomUUID().toString())
```
Comment (Flowdalic) [inline smack-extensions/src/test/java/org/jivesoftware/smackx/muc/MultiUserChatTest.java:121]:
```suggestion
        Message sentMessage = connection.getStanzaFactory().buildMessage()
```

File: smack-extensions/src/test/java/org/jivesoftware/smackx/muc/MultiUserChatTest.java [hunk 12]
```diff
@@ -0,0 +1,211 @@
+/**
+ *
+ * Copyright 2021 Microsoft Corporation.
```
Comment (Flowdalic) [inline smack-extensions/src/test/java/org/jivesoftware/smackx/muc/MultiUserChatTest.java:3]: Correct date? Should probably be at least 2021-2022.

File: smack-extensions/src/test/java/org/jivesoftware/smackx/muc/MultiUserChatTest.java [hunk 13]
```diff
@@ -0,0 +1,211 @@
...
+        assertTrue(updateRequestSent.get(), "Invite directly request not sent");
+    }
+
+    @Test
+    public void shouldReceiveOfflineInvitation() throws XmppStringprepException {
+        EntityBareJid roomJid = JidCreate.entityBareFrom("room@example.com");
+        EntityFullJid inviterJid = JidCreate.entityFullFrom("inviter@example.com/user1");
+        EntityBareJid inviteeJid = JidCreate.entityBareFrom("invitee@example.com");
+
+        GroupChatInvitation groupChatInvitation = new GroupChatInvitation(JidCreate.entityBareFrom("room@example.com"));
```
Comment (Flowdalic) [inline smack-extensions/src/test/java/org/jivesoftware/smackx/muc/MultiUserChatTest.java:120]:
```suggestion
        GroupChatInvitation groupChatInvitation = new GroupChatInvitation(roomJid);
```

File: smack-extensions/src/test/java/org/jivesoftware/smackx/muc/MultiUserChatTest.java [hunk 14]
```diff
@@ -0,0 +1,211 @@
...
+                .addExtension(groupChatInvitation)
+                .build();
+
+        // Prepare listener to receive a group invitation
+        GroupInvitationListener groupInvitationListener = new GroupInvitationListener() {
+            @Override
+            public void verifyInvitation(XMPPConnection conn, MultiUserChat room, EntityJid inviter,
+                                         String reason, String password, Message message, MUCUser.Invite invitation) {
+                // Check all parameters values.
+                assertSame(connection, conn);
```
Comment (Flowdalic) [inline smack-extensions/src/test/java/org/jivesoftware/smackx/muc/MultiUserChatTest.java:133]: I fear those asserts won't work, as their exceptions will be not be thrown in the thread that invoked the unit test method.

File: smack-extensions/src/test/java/org/jivesoftware/smackx/muc/MultiUserChatTest.java [hunk 15]
```diff
@@ -0,0 +1,211 @@
...
+
+        // Simulate sending a message with a group invitation
+        connection.processStanza(sentMessage);
+
+        // Wait for the listener to be called or throw a timeout exception
+        groupInvitationListener.waitUntilInvocationOrTimeout();
+    }
+
+    private static class Responder extends Thread {
+        protected final AtomicBoolean running = new AtomicBoolean();
```
Comment (Flowdalic) [inline smack-extensions/src/test/java/org/jivesoftware/smackx/muc/MultiUserChatTest.java:153]: This field isn't used anywhere, in fact the whole `Responder` class could be merged into `InvokeDirectlyResponder`. I assume this was potentially at some point in the past different, but right now it is not required in this form and hence only contributes to confusion, i.e., it makes the code harder to read.

**Review by Flowdalic — CHANGES_REQUESTED** [review]

File: smack-extensions/src/main/java/org/jivesoftware/smackx/muc/MultiUserChatManager.java (hunk 4 above)
[the remark this answers is not in the document]
Comment (Flowdalic) [reply smack-extensions/src/main/java/org/jivesoftware/smackx/muc/MultiUserChatManager.java]: My fault. I wasn't looking carefully enough and assumed this is in `MultiUserChat`.

File: smack-extensions/src/main/java/org/jivesoftware/smackx/muc/MultiUserChatManager.java (hunk 8 above)
Comment (MF1-MS) [reply smack-extensions/src/main/java/org/jivesoftware/smackx/muc/MultiUserChatManager.java:230]: I'm not sure I agree. The end user the invitation is from should be the same regardless of if they're using an XEP-0249 direct invitation or a mediated XEP-0045 invitation. If we are asserting a user must be an entity jid in the XEP-0045 case then I think it's very reasonable to assert the same here.

File: smack-extensions/src/test/java/org/jivesoftware/smackx/muc/MultiUserChatTest.java (hunk 14 above)
Comment (MF1-MS) [reply smack-extensions/src/test/java/org/jivesoftware/smackx/muc/MultiUserChatTest.java:133]: Hmmm you're right, it throws by then we have to wait for the latch to time out, and then we can check the stderr in the test report to diagnose the problem. Any suggestions?

```
> Task :smack-extensions:test

MultiUserChatTest > shouldReceiveOfflineInvitationWithAllOptionalAttributes() FAILED
    java.lang.IllegalStateException: Latch timed out before it reached zero
        at org.jivesoftware.smack.test.util.WaitForPacketListener.waitUntilInvocationOrTimeout(WaitForPacketListener.java:49)
        at org.jivesoftware.smackx.muc.MultiUserChatTest.shouldReceiveOfflineInvitationWithAllOptionalAttributes(MultiUserChatTest.java:248)
```

File: smack-extensions/src/main/java/org/jivesoftware/smackx/muc/MultiUserChat.java (hunk 5 above)
[the remark this answers is not in the document]
Comment (MF1-MS) [reply smack-extensions/src/main/java/org/jivesoftware/smackx/muc/MultiUserChat.java]: Done.

File: smack-extensions/src/test/java/org/jivesoftware/smackx/muc/MultiUserChatTest.java (hunk 14 above)
Comment (MF1-MS) [reply smack-extensions/src/test/java/org/jivesoftware/smackx/muc/MultiUserChatTest.java:183]: Ah, I can have it save off an exception we check and rethrow in the main thread like the `InvokeDirectlyResponder`

**Comment by MF1-MS** [comment]
OK sorry about the delay, that's amended now. I can squash the commits and amend the message if you're happy, but leaving as is for now for a cleaner diff. 

Are you happy with my other PRs too?:
https://github.com/igniterealtime/Smack/pull/537
https://github.com/igniterealtime/Smack/pull/538

File: build.gradle [hunk 16]
```diff
@@ -316,7 +316,7 @@ tasks.withType(Javadoc) {
 		// fixtures, and we want to have mockito also available in
 		// test, so we use API here.
 		testFixturesApi "org.mockito:mockito-core:${mockitoVersion}"
-
```
Comment (Flowdalic) [inline build.gradle:319]: Unrelated whitespace change, please remove this from the PR (but feel encouraged to add it as new PR).

File: smack-core/src/testFixtures/java/org/jivesoftware/smack/test/util/WaitForPacketListener.java [hunk 17]
```diff
@@ -43,7 +43,7 @@ public void waitAndReset() {
 
     public void waitUntilInvocationOrTimeout() {
         try {
-            boolean res = latch.await(300, TimeUnit.SECONDS);
+            boolean res = latch.await(10, TimeUnit.SECONDS);
```
Comment (Flowdalic) [inline smack-core/src/testFixtures/java/org/jivesoftware/smack/test/util/WaitForPacketListener.java:46]: I don't see a reason why this should be lowered to 10 seconds. Especially since 10 seconds is easily in the area where heavily congested systems may delay a thread. I'd like to prefer the original value, we could, however talk about making this configurable. But then, as above, no as part of this PR.

File: smack-extensions/src/main/java/org/jivesoftware/smackx/muc/MultiUserChat.java [hunk 18]
```diff
@@ -1063,6 +1064,52 @@ public void invite(MessageBuilder messageBuilder, EntityBareJid user, String rea
...
+     *
+     * @param address the user to send the invitation to
+     * @param reason the purpose for the invitation
+     * @param password specifies a password needed for entry
+     * @param continue_bool specifies if the groupchat room continues a one-to-one chat having the designated thread
+     * @param thread the thread to continue
+     * @throws NotConnectedException if the XMPP connection is not connected.
+     * @throws InterruptedException if the calling thread was interrupted.
+     */
+    public void inviteDirectly(EntityBareJid address, String reason, String password, boolean continue_bool, String thread)
```
Comment (Flowdalic) [inline smack-extensions/src/main/java/org/jivesoftware/smackx/muc/MultiUserChat.java:1095]: Smack uses camel case, as most of the Java ecosystem, not snake case. I also think we do not need to encode the type in the parameter name.
```suggestion
    public void inviteDirectly(EntityBareJid address, String reason, String password, boolean continue, String thread)
```

File: smack-extensions/src/main/java/org/jivesoftware/smackx/muc/MultiUserChat.java [hunk 19]
```diff
@@ -1063,6 +1064,52 @@ public void invite(MessageBuilder messageBuilder, EntityBareJid user, String rea
...
+        throws NotConnectedException, InterruptedException {
+        MessageBuilder messageBuilder = connection.getStanzaFactory().buildMessageStanza();
```
Comment (Flowdalic) [inline smack-extensions/src/main/java/org/jivesoftware/smackx/muc/MultiUserChat.java:1097]: Use the builder pattern if it is available
```suggestion
        Message message = connection.getStanzaFactory()
            .buildMessageStanza()
            .to(address)
            .addExtension(invitation)
            .build();
```

File: smack-extensions/src/main/java/org/jivesoftware/smackx/muc/MultiUserChatManager.java [hunk 20]
```diff
@@ -199,6 +206,33 @@ public void processStanza(Stanza packet) {
         };
         connection.addAsyncStanzaListener(invitationPacketListener, INVITATION_FILTER);
 
+        // Listens for all messages that include an XEP-0249 GroupChatInvitation extension and fire the invitation
+        // listeners
+        StanzaListener directInvitationPacketListener = new StanzaListener() {
+            @Override
+            public void processStanza(Stanza packet) {
```
Comment (Flowdalic) [inline smack-extensions/src/main/java/org/jivesoftware/smackx/muc/MultiUserChatManager.java:213]: We try to phase out the term `packet` in Smack, as it is not XMPP-ish.
```suggestion
            public void processStanza(Stanza stanza) {
```

File: smack-extensions/src/main/java/org/jivesoftware/smackx/muc/packet/GroupChatInvitation.java [hunk 21]
```diff
@@ -79,7 +83,67 @@ public class GroupChatInvitation implements ExtensionElement {
...
+     * @param roomAddress the address of the group chat room.
+     * @param reason the purpose for the invitation
+     * @param password specifies a password needed for entry
+     * @param continue_bool specifies if the groupchat room continues a one-to-one chat having the designated thread
+     * @param thread the thread to continue
+     */
+    public GroupChatInvitation(EntityBareJid roomAddress,
+                               String reason,
+                               String password,
+                               Boolean continue_bool,
```
Comment (Flowdalic) [inline smack-extensions/src/main/java/org/jivesoftware/smackx/muc/packet/GroupChatInvitation.java:104]: No reason to use a boxed type, also see above regarding snake case.
```suggestion
                               boolean continue,
```

File: smack-extensions/src/main/java/org/jivesoftware/smackx/muc/packet/GroupChatInvitation.java [hunk 22]
```diff
@@ -69,6 +69,10 @@ public class GroupChatInvitation implements ExtensionElement {
     public static final QName QNAME = new QName(NAMESPACE, ELEMENT);
 
     private final EntityBareJid roomAddress;
+    private final String reason;
+    private final String password;
+    private final String thread;
+    private final Boolean continue_bool;
```
Comment (Flowdalic) [inline smack-extensions/src/main/java/org/jivesoftware/smackx/muc/packet/GroupChatInvitation.java:75]: No reason to use a boxed type, also see above regarding snake case.
```suggestion
    private final boolean continue;
```

File: smack-extensions/src/main/java/org/jivesoftware/smackx/muc/packet/GroupChatInvitation.java [hunk 23]
```diff
@@ -79,7 +83,67 @@ public class GroupChatInvitation implements ExtensionElement {
...
+    public String getThread() {
+        return thread;
+    }
+
+    /**
+     * Returns whether the groupchat room continues a one-to-one chat.
+     *
+     * @return whether the groupchat room continues a one-to-one chat.
+     */
+    public Boolean getContinue() {
```
Comment (Flowdalic) [inline smack-extensions/src/main/java/org/jivesoftware/smackx/muc/packet/GroupChatInvitation.java:145]: I am not really happy about the method name, as it is not expressive. Also return type. Maybe
```suggestion
    public boolean continuesAsOneToOneChat() {
```

File: smack-extensions/src/main/java/org/jivesoftware/smackx/muc/packet/GroupChatInvitation.java [hunk 24]
```diff
@@ -107,6 +171,13 @@ public String getNamespace() {
     public XmlStringBuilder toXML(org.jivesoftware.smack.packet.XmlEnvironment enclosingNamespace) {
         XmlStringBuilder xml = new XmlStringBuilder(this);
         xml.attribute("jid", getRoomAddress());
+        xml.optAttribute("reason", getReason());
+        xml.optAttribute("password", getPassword());
+        xml.optAttribute("thread", getThread());
+
+        if (getContinue())
+            xml.optAttribute("continue", getContinue().toString());
```
Comment (Flowdalic) [inline smack-extensions/src/main/java/org/jivesoftware/smackx/muc/packet/GroupChatInvitation.java:179]: I think `XmlStringBuilder` provides `optBoolAttribute()`.

File: smack-extensions/src/test/java/org/jivesoftware/smackx/muc/MultiUserChatTest.java [hunk 25]
```diff
@@ -0,0 +1,324 @@
...
+
+        if (groupInvitationListener.getError() != null) {
+            throw groupInvitationListener.getError();
+        }
+    }
+
+    @Test
+    public void shouldReceiveOfflineInvitationWithAllOptionalAttributes() throws XmppStringprepException, Throwable {
+        EntityBareJid roomJid = JidCreate.entityBareFrom("room@example.com");
+        String expected_reason = "reason";
```
Comment (Flowdalic) [inline smack-extensions/src/test/java/org/jivesoftware/smackx/muc/MultiUserChatTest.java:216]: camel case, please

File: smack-extensions/src/main/java/org/jivesoftware/smackx/muc/MultiUserChat.java (hunk 18 above)
Comment (MF1-MS) [reply smack-extensions/src/main/java/org/jivesoftware/smackx/muc/MultiUserChat.java:1095]: `continue` is a Java keyword, so I'd rather not just call it that. Have amended to `continueAsOneToOne`

Comment (Flowdalic) [reply smack-extensions/src/main/java/org/jivesoftware/smackx/muc/MultiUserChat.java:1095]: +1, but maybe go all the way and use `continueAsOneToOneChat`.

Comment (MF1-MS) [reply smack-extensions/src/main/java/org/jivesoftware/smackx/muc/MultiUserChat.java:1095]: Yeh I'd gone back and done that on second thoughts 👍

**Comment by MF1-MS** [comment]
> I am not sure if you re-invented `org.jivesoftware.util.Protocol` with `InvokeDirectlyResponder` in your tests (or similar already existing classes of Smack's test fixtures). Please check that you don't invent the wheel again, I wouldn't be surprised if Smack wouldn't already provide what you need for the tests. Unfortunately, I had no time to dig into this further.

I think these are suitably disparate. `org.jivesoftware.util.Protocol` (which seems to have an out of data docstring, but I found good usage examples in the Socks5 tests) seems focussed on enabling dialogues where Smack will be expecting responses to its requests, and verifying the requests Smack sent. The sent stanza we're verifying with `InvokeDirectlyResponder` is generated by Smack due to it conceptually receiving a message. 

`InvokeDirectlyResponder` simply builds off the `DummyConnection` test fixture, capturing an outgoing stanza and verifying it.

**Comment by MF1-MS** [comment]
I can squash the commits and amend the message if you're happy, but leaving as is for now for a cleaner diff.

Are you happy with my other PRs too?:
https://github.com/igniterealtime/Smack/pull/537
https://github.com/igniterealtime/Smack/pull/538

File: smack-extensions/src/main/java/org/jivesoftware/smackx/muc/MultiUserChatManager.java [hunk 26]
```diff
@@ -199,6 +206,33 @@ public void processStanza(Stanza packet) {
...
+                final Message message = (Message) stanza;
+                GroupChatInvitation invite =
+                    stanza.getExtension(GroupChatInvitation.class);
+
+                // Fire event for invitation listeners
+                final MultiUserChat muc = getMultiUserChat(invite.getRoomAddress());
+                final XMPPConnection connection = connection();
+                final EntityJid from = message.getFrom().asEntityJidIfPossible();
+                if (from == null) {
+                    LOGGER.warning("Group Chat Invitation from non entity JID: '" + message.toXML() + "'");
```
Comment (Flowdalic) [inline smack-extensions/src/main/java/org/jivesoftware/smackx/muc/MultiUserChatManager.java:223]: Do not spit out the whole XML of the causing stanza if it is not a debug message.
```suggestion
                    LOGGER.warning("Group Chat Invitation from non entity JID in '" + message + "'");
```

**Review by Flowdalic — CHANGES_REQUESTED** [review]
LGTM, only one minor thing, then this is ready to go.

Please
1. squash the commits into one
2. Add an table row for xep249 to https://github.com/igniterealtime/Smack/blob/f6c85d9fb3a4d8baa96d48dc428b53974d9fdba5/smack-java8-full/src/main/java/org/jivesoftware/smackx/package-info.java#L348-L361
3. rebase on the current master
4.  mention "Fixes SMACK-932" in the commit message's body. Also

**Comment by MF1-MS** [comment]
Should be good to merge now :)

**Comment by Flowdalic** [comment]
The commit message's subject currently reads "Fixes SMACK-932.". That is not really helpful if I look at a commit log only consisting of the commit subject.

**Comment by MF1-MS** [comment]
Ok, commit message re-arranged.
